### DISCUSSION TITLE
The event viewer lacks description of nssm events when using ServiceInstall

### DISCUSSION THREAD

**Discussion by marbx** [wixtoolset/issues#6289] [Questions] [answered]
An application uses [nssm](https://nssm.cc) as services manager.
When installed with WiX ServiceInstall, the service starts, stops and works, but the event viewer contains for all events #### from source nssm: 

    The description for Event ID #### from source nssm cannot be found. 
    Either the component that raises this event is not installed on your local computer or the installation is corrupted. 

What is the reason and how to fix it?

**Comment by wyrdfish** [comment]
Presumably nssm registers itself as an event source if you use it to install the service.
But if you are using wix to do the service installation it will not know about nssm. so you will need to register them manually.

https://stackoverflow.com/questions/58538/how-do-you-create-an-event-log-source-using-wix

**Reply by marbx** [reply]
I get it, but I tried using `nssm` and the name of the service as `util:EventSource@Name`,  and neither worked.
In Github I found many examples with a *.dll as `util:EventSource@EventMessageFile` and only a handful with *.exe

What am I doing wrong?

```
        <util:EventSource
          Log="Application"
          Name="nssm"
          EventMessageFile="[#nssm.exe]" />
```

**Reply by wyrdfish** [reply]
I don't actually use wix to do this but I do use nssm from chef.

under HKLM\System\CurrentControlSet\Services\EventLog\Application\NSSM i have entries.
EventMessageFile is a full path to the nssm exe
there is also a TypesSupported dword with value 7.

Do you have anything there?

I also read that you may need to reboot to get event log registration to work.

**Reply by marbx** [reply]
Yes, but I have only one entry under `HKLM\System\CurrentControlSet\Services\EventLog\Application\NSSM`: 
- `TypesSupported` (dword with value 7)
- (`EventMessageFile` is missing)

You execute nssm (from chef). As you wrote and as I can confirm, nssm registers itself as an event source.

For the WiX part:
- `ServiceInstall` registers nssm.exe as a service, but it probably does not register it as an event listener (for itself) "at the same time", and why should it.
- `util:EventSource` (which you don't need nor use) probably does not or cannot register nssm.exe as an event log application either.

If both is correct: is it possible to register nssm as event log application (without running nssm.exe in a WixQuietExec custom action)?

**Reply by wyrdfish** [reply]
Is the <util:EventSource> not working then?
check a log to see what parameters you are passing to it.
maybe resolve the path to nssm in some other way if [#nssm.exe] is not working?
do you have a file with id nssm.exe in your wix?

**Comment by marbx** [comment]
A little progress.

> Is the util:EventSource not working then?

Unfortunately not, all events start with "The description for Event ID #### from source nssm cannot be found."


Even though I now find 2 entries under HKLM\System\CurrentControlSet\Services\EventLog\Application\NSSM:
- TypesSupported (dword with value 7)
- EventMessageFile (C:\salt\ssm.exe)

Maybe because I added `CreateFolder` [as suggested here](https://stackoverflow.com/a/14451992/3173832) 

ssm.exe  is not a typo, saltstack compiled it's own nssm and left out the "n" in the file name (not in the event source)
Installed with a Nullsoft, this ssm.exe creates regular Windows events.

I see no difference between the registry settings after Nullsoft and Wix installs.

I have not tried to reboot as this is not an option.


> check a log to see what parameters you are passing to it.

Do you mean this log?

```
MSI (s) (74:60) [19:02:18:310]: Executing op: ActionStart(Name=WriteRegistryValues,Description=Writing system registry values,Template=Key: [1], Name: [2], Value: [3])
MSI (s) (74:60) [19:02:18:312]: Executing op: ProgressTotal(Total=9,Type=1,ByteEquivalent=13200)
MSI (s) (74:60) [19:02:18:313]: Executing op: RegOpenKey(Root=-2147483646,Key=SYSTEM\CurrentControlSet\Services\EventLog\Application\nssm,,BinaryType=1,,)
MSI (s) (74:60) [19:02:18:313]: Executing op: RegAddValue(Name=EventMessageFile,Value=#%C:\salt\ssm.exe,)
MSI (s) (74:60) [19:02:18:315]: Executing op: RegOpenKey(Root=-2147483646,Key=System\CurrentControlSet\Services\salt-minion\Parameters,,BinaryType=1,,)
MSI (s) (74:60) [19:02:18:316]: Executing op: RegAddValue(Name=AppDirectory,Value=#%C:\salt\bin,)
MSI (s) (74:60) [19:02:18:320]: Executing op: RegAddValue(Name=Application,Value=#%C:\salt\bin\python.exe,)
MSI (s) (74:60) [19:02:18:321]: Executing op: RegAddValue(Name=AppParameters,Value=#%-E -s C:\salt\bin\Scripts\salt-minion -c C:\salt\conf -l quiet,)
MSI (s) (74:60) [19:02:18:322]: Executing op: RegAddValue(Name=AppStopMethodConsole,Value=#24000,)
MSI (s) (74:60) [19:02:18:323]: Executing op: RegAddValue(Name=AppStopMethodWindow,Value=#2000,)
MSI (s) (74:60) [19:02:18:324]: Executing op: RegAddValue(Name=AppRestartDelay,Value=#60000,)
MSI (s) (74:60) [19:02:18:325]: Executing op: RegOpenKey(Root=-2147483646,Key=System\CurrentControlSet\Services\salt-minion\Parameters\AppExit,,BinaryType=1,,)
MSI (s) (74:60) [19:02:18:325]: Executing op: RegAddValue(,Value=Restart,)
MSI (s) (74:60) [19:02:18:326]: Executing op: RegOpenKey(Root=-2147483646,Key=SOFTWARE\SaltStack\Salt-Minion,,BinaryType=1,,)
MSI (s) (74:60) [19:02:18:326]: Executing op: RegAddValue(Name=RememberForRemoveFolderExBINFOLDER,Value=C:\salt\bin\,)
```


> maybe resolve the path to nssm in some other way if [#nssm.exe] is not working?

No need, as the file is found.


>do you have a file with id nssm.exe in your wix?

Yes 


Question remains: is it possible to register nssm as event log application?

**Reply by robmen** [reply]
> Question remains: is it possible to register nssm as event log application?

Seems like that would be a question better asked for the nssm team. It absolutely is possible to register an executable with the event log.

**Reply by marbx** [reply]
I may have sounded rude and apologize. 

I would like to understand more about the event log, and what I am doing wrong.

**Reply by robmen** [reply]
The event log is defined by [a bunch of registry keys](https://docs.microsoft.com/en-us/windows/win32/eventlog/eventlog-key). The specifics of those keys are dependent on the application you are registering. I wasn't trying to be flippant when I said:

> Seems like that would be a question better asked for the nssm team.

**Comment by marbx** [comment]
Thank you both.
I will continue reading and come back.

**Comment by marbx** [comment, marked as answer]
I created a very small Python program for testing.

I noted that after install:
- The installed application works.
- The nssm event log contains readable messages like `Service nssm-test-loop received START control, which will be handled.`

This proves that WiX can install nssm as a service.

After uninstall:
- All nssm event log entries now start with the warning "The description for Event ID 1023 from source nssm cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer."

I assume this is perfectly expected behavior. 

After installation of the same msi again:
- The installed application works.
- All nssm event logs still start with the warning.

This surprises me.

After building the installer with a new component GUID (at ABC123 below) and install, without reboot:
- The installed application works.
- All nssm event log messages are readable again.

I assume that the underlying problem is nssm, which is unsupported. 
I will therefore close my question, unless you can point me to a mistake in my code below.


```
<?xml version="1.0" encoding="utf-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi"
  xmlns:util="http://schemas.microsoft.com/wix/UtilExtension">
  <Fragment>
    <ComponentGroup Id="service">
      <Component Directory="TARGETDIR" Guid="{ABC123}">
        <File Id="nssm.exe" KeyPath="yes" Source="....\nssm.exe" />
        <ServiceInstall
          Account="LocalSystem"
          ErrorControl="normal"
          Name="nssm-test-loop"
          Start="auto"
          Type="ownProcess"
          Vital="yes"
          Description="nssm-test-loop"
          DisplayName="nssm-test-loop"
          Id="nssm-test-loop">
          <util:ServiceConfig
            FirstFailureActionType="none"
            SecondFailureActionType="none"
            ThirdFailureActionType="none" />
        </ServiceInstall>
        <ServiceControl
          Id="nssm-test-loop"
          Name="nssm-test-loop"
          Remove="uninstall"
          Stop="both"
          Start="install"
          Wait="yes">
          <ServiceArgument />
        </ServiceControl>
        <CreateFolder />
        <util:EventSource
          Log="Application"
          Name="nssm"
          EventMessageFile="[#nssm.exe]" />
        <RegistryKey Root="HKLM" Key="System\CurrentControlSet\Services\nssm-test-loop">
          <RegistryKey Key="Parameters">
            <RegistryValue Type="expandable" Name="AppDirectory"         Value="c:\Python3" />
            <RegistryValue Type="expandable" Name="Application"          Value="c:\Python3\python.exe" />
            <RegistryValue Type="expandable" Name="AppParameters"        Value="....\test_loop.py" />
            <RegistryValue Type="integer"    Name="AppStopMethodConsole" Value="24000" />
            <RegistryValue Type="integer"    Name="AppStopMethodWindow"  Value="2000" />
            <RegistryValue Type="integer"    Name="AppRestartDelay"      Value="60000" />
            <RegistryKey Key="AppExit">
              <RegistryValue Type="string" Value="Restart" />
            </RegistryKey>
          </RegistryKey>
        </RegistryKey>
      </Component>
    </ComponentGroup>
  </Fragment>
</Wix>

```
